README.txt
Download and docs:
    http://pypi.python.org/pypi/colorama
Development:
    http://code.google.com/p/colorama
Discussion group:
     https://groups.google.com/forum/#!forum/python-colorama

Description
===========

Makes ANSI escape character sequences for producing colored terminal text and
cursor positioning work under MS Windows.

ANSI escape character sequences have long been used to produce colored terminal
text and cursor positioning on Unix and Macs. Colorama makes this work on
Windows, too, by wrapping stdout, stripping ANSI sequences it finds (which
otherwise show up as gobbledygook in your output), and converting them into the
appropriate win32 calls to modify the state of the terminal. On other platforms,
Colorama does nothing.

Colorama also provides some shortcuts to help generate ANSI sequences
but works fine in conjunction with any other ANSI sequence generation library,
such as Termcolor (http://pypi.python.org/pypi/termcolor.)

This has the upshot of providing a simple cross-platform API for printing
colored terminal text from Python, and has the happy side-effect that existing
applications or libraries which use ANSI sequences to produce colored output on
Linux or Macs can now also work on Windows, simply by calling
``colorama.init()``.

An alternative approach is to install 'ansi.sys' on Windows machines, which
provides the same behaviour for all applications running in terminals. Colorama
is intended for situations where that isn't easy (e.g. maybe your app doesn't
have an installer.)

Demo scripts in the source code repository prints some colored text using
ANSI sequences. Compare their output under Gnome-terminal's built in ANSI
handling, versus on Windows Command-Prompt using Colorama:

.. image:: http://colorama.googlecode.com/hg/screenshots/ubuntu-demo.png
    :width: 661
    :height: 357
    :alt: ANSI sequences on Ubuntu under gnome-terminal.

.. image:: http://colorama.googlecode.com/hg/screenshots/windows-demo.png
    :width: 668
    :height: 325
    :alt: Same ANSI sequences on Windows, using Colorama.

These screengrabs show that Colorama on Windows does not support ANSI 'dim
text': it looks the same as 'normal text'.


License
=======

Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.


Dependencies
============

None, other than Python. Tested on Python 2.5.5, 2.6.5, 2.7, 3.1.2, and 3.2

Usage
=====

Initialisation
--------------

Applications should initialise Colorama using::

    from colorama import init
    init()

If you are on Windows, the call to ``init()`` will start filtering ANSI escape
sequences out of any text sent to stdout or stderr, and will replace them with
equivalent Win32 calls.

Calling ``init()`` has no effect on other platforms (unless you request other
optional functionality, see keyword args below.) The intention is that
applications can call ``init()`` unconditionally on all platforms, after which
ANSI output should just work.

To stop using colorama before your program exits, simply call ``deinit()``.
This will restore stdout and stderr to their original values, so that Colorama
is disabled. To start using Colorama again, call ``reinit()``, which wraps
stdout and stderr again, but is cheaper to call than doing ``init()`` all over
again.


Colored Output
--------------

Cross-platform printing of colored text can then be done using Colorama's
constant shorthand for ANSI escape sequences::

    from colorama import Fore, Back, Style
    print(Fore.RED + 'some red text')
    print(Back.GREEN + 'and with a green background')
    print(Style.DIM + 'and in dim text')
    print(Fore.RESET + Back.RESET + Style.RESET_ALL)
    print('back to normal now')

or simply by manually printing ANSI sequences from your own code::

    print('/033[31m' + 'some red text')
    print('/033[30m' # and reset to default color)

or Colorama can be used happily in conjunction with existing ANSI libraries
such as Termcolor::

    from colorama import init
    from termcolor import colored

    # use Colorama to make Termcolor work on Windows too
    init()

    # then use Termcolor for all colored text output
    print(colored('Hello, World!', 'green', 'on_red'))

Available formatting constants are::

    Fore: BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, WHITE, RESET.
    Back: BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, WHITE, RESET.
    Style: DIM, NORMAL, BRIGHT, RESET_ALL

Style.RESET_ALL resets foreground, background and brightness. Colorama will
perform this reset automatically on program exit.


Cursor Positioning
------------------

ANSI codes to reposition the cursor are supported. See demos/demo06.py for
an example of how to generate them.


Init Keyword Args
-----------------

``init()`` accepts some kwargs to override default behaviour.

init(autoreset=False):
    If you find yourself repeatedly sending reset sequences to turn off color
    changes at the end of every print, then ``init(autoreset=True)`` will
    automate that::

        from colorama import init
        init(autoreset=True)
        print(Fore.RED + 'some red text')
        print('automatically back to default color again')

init(strip=None):
    Pass ``True`` or ``False`` to override whether ansi codes should be
    stripped from the output. The default behaviour is to strip if on Windows.

init(convert=None):
    Pass ``True`` or ``False`` to override whether to convert ansi codes in the
    output into win32 calls. The default behaviour is to convert if on Windows
    and output is to a tty (terminal).

init(wrap=True):
    On Windows, colorama works by replacing ``sys.stdout`` and ``sys.stderr``
    with proxy objects, which override the .write() method to do their work. If
    this wrapping causes you problems, then this can be disabled by passing
    ``init(wrap=False)``. The default behaviour is to wrap if autoreset or
    strip or convert are True.

    When wrapping is disabled, colored printing on non-Windows platforms will
    continue to work as normal. To do cross-platform colored output, you can
    use Colorama's ``AnsiToWin32`` proxy directly::

        import sys
        from colorama import init, AnsiToWin32
        init(wrap=False)
        stream = AnsiToWin32(sys.stderr).stream

        # Python 2
        print >>stream, Fore.BLUE + 'blue text on stderr'    

        # Python 3
        print(Fore.BLUE + 'blue text on stderr', file=stream)
        

Status & Known Problems
=======================

I've personally only tested it on WinXP (CMD, Console2), Ubuntu 
(gnome-terminal, xterm), and OSX.

Some presumably valid ANSI sequences aren't recognised (see details below)
but to my knowledge nobody has yet complained about this. Puzzling.

See outstanding issues and wishlist at:
http://code.google.com/p/colorama/issues/list

If anything doesn't work for you, or doesn't do what you expected or hoped for,
I'd love to hear about it on that issues list, would be delighted by patches,
and would be happy to grant commit access to anyone who submits a working patch
or two.


Recognised ANSI Sequences
=========================

ANSI sequences generally take the form:

    ESC [ <param> ; <param> ... <command>

Where <param> is an integer, and <command> is a single letter. Zero or more 
params are passed to a <command>. If no params are passed, it is generally
synonymous with passing a single zero. No spaces exist in the sequence, they
have just been inserted here to make it easy to read.

The only ANSI sequences that colorama converts into win32 calls are::

    ESC [ 0 m       # reset all (colors and brightness)
    ESC [ 1 m       # bright
    ESC [ 2 m       # dim (looks same as normal brightness)
    ESC [ 22 m      # normal brightness

    # FOREGROUND:
    ESC [ 30 m      # black
    ESC [ 31 m      # red
    ESC [ 32 m      # green
    ESC [ 33 m      # yellow
    ESC [ 34 m      # blue
    ESC [ 35 m      # magenta
    ESC [ 36 m      # cyan
    ESC [ 37 m      # white
    ESC [ 39 m      # reset

    # BACKGROUND
    ESC [ 40 m      # black
    ESC [ 41 m      # red
    ESC [ 42 m      # green
    ESC [ 43 m      # yellow
    ESC [ 44 m      # blue
    ESC [ 45 m      # magenta
    ESC [ 46 m      # cyan
    ESC [ 47 m      # white
    ESC [ 49 m      # reset

    # cursor positioning
    ESC [ y;x H     # position cursor at x across, y down

    # clear the screen
    ESC [ mode J    # clear the screen. Only mode 2 (clear entire screen)
                    # is supported. It should be easy to add other modes,
                    # let me know if that would be useful.

Multiple numeric params to the 'm' command can be combined into a single
sequence, eg::

    ESC [ 36 ; 45 ; 1 m     # bright cyan text on magenta background

All other ANSI sequences of the form ``ESC [ <param> ; <param> ... <command>``
are silently stripped from the output on Windows.

Any other form of ANSI sequence, such as single-character codes or alternative
initial characters, are not recognised nor stripped. It would be cool to add
them though. Let me know if it would be useful for you, via the issues on
google code.


Development
===========

Running tests requires:

- Michael Foord's 'mock' module to be installed.
- Tests are written using the 2010 era updates to 'unittest', and require to
  be run either using Python2.7 or greater, or else to have Michael Foord's
  'unittest2' module installed.

unittest2 test discovery doesn't work for colorama, so I use 'nose'::

    nosetests -s

The -s is required because 'nosetests' otherwise applies a proxy of its own to
stdout, which confuses the unit tests.


Thanks
======
| Jesse@EmptySquare for submitting a fix for examples in the README.
| User 'jamessp', an observant documentation fix for cursor positioning.
| User 'vaal1239', Dave Mckee & Lackner Kristof for a tiny but much-needed Win7 fix.
| Julien Stuyck, for wisely suggesting Python3 compatible updates to README.
| Daniel Griffith for multiple fabulous patches.
| Oscar Lesta for valuable fix to stop ANSI chars being sent to non-tty output.
| Roger Binns, for many suggestions, valuable feedback, & bug reports.
| Tim Golden for thought and much appreciated feedback on the initial idea.



CHANGELOG.rst
0.2.6
    Add copyright & licensing info to every file, as requested by a large
    downstream project which has problems making sure that all 3rd party
    contributions have appropriate license.
0.2.5
    Severeral documentation & demo fixes.
0.2.4
    Fix to work on Windows 7.
    Python 3 compatibility in docs and demos.
    Add handling for 'cursor up' and 'get position' ANSI codes.
0.2.3
	Split changelog out into separate file.
0.2.2
    Fix bug which caused init() to raise, introduced in 0.2.1.
    Remove asserts which cause problems in various circumstances. At least
    some users saw asserts fail on 'success' returned from win32 functions,
    even though the win32 functions appear to have worked correctly.
0.2.1
    Completely broken: I added a bug which caused init() to raise.
    Added some documentation for cursor positioning and clear screen to README.
    Add 'reinit' and 'deinit' functions, as suggested by Charles FOL and
    Romanov DA.
0.2
    Merge in changes from Daniel Griffith: Add ANSI cursor positioning &
    partial support for clear screen. Patch submitted by Oscar Lester, don't
    send RESET_ALL to non-tty. Demos split into separate files and moved into
    their own directory. Tweak sys.path in demos so they run against local
    source, not installed version of Colorama.
0.1.18
    Fix README (no such attr as Fore.DEFAULT, etc), kindly reported by nodakai.
0.1.17
    Prevent printing of garbage ANSI codes upon installing with pip
0.1.16
    Re-upload to fix previous error. Make clean now removes old MANIFEST.
0.1.15
    Completely broken. Distribution was empty due to leftover invalid MANIFEST
    file from building on a different platform.
    Fix python3 incompatibility kindly reported by G |uumlaut| nter Kolousek
0.1.14
    Fix hard-coded reset to white-on-black colors. Fore.RESET, Back.RESET
    and Style.RESET_ALL now revert to the colors as they were when init()
    was called. Some lessons hopefully learned about testing prior to release.
0.1.13
    Completely broken: barfed when installed using pip.
0.1.12
    Completely broken: contained no source code. double oops.
0.1.11
    Completely broken: fatal import errors on Ubuntu. oops.
0.1.10
    Stop emulating 'bright' text with bright backgrounds.
    Display 'normal' text using win32 normal foreground instead of bright.
    Drop support for 'dim' text.
0.1.9
    Fix incompatibility with Python 2.5 and earlier.
    Remove setup.py dependency on setuptools, now uses stdlib distutils.
0.1.8
    Fix ghastly errors all over the place on Ubuntu.
    Add init kwargs 'convert' and 'strip', which supercede the old 'wrap'.
0.1.7
    Python 3 compatible.
    Fix: Now strips ansi on windows without necessarily converting it to
    win32 calls (eg. if output is not a tty.)
    Fix: Flaky interaction of interleaved ansi sent to stdout and stderr.
    Improved demo.sh (hg checkout only.)
0.1.6
    Fix ansi sequences with no params now default to parmlist of [0].
    Fix flaky behaviour of autoreset and reset_all atexit.
    Fix stacking of repeated atexit calls - now just called once.
    Fix ghastly import problems while running tests.
    'demo.py' (hg checkout only) now demonstrates autoreset and reset atexit.
    Provide colorama.VERSION, used by setup.py.
    Tests defanged so they no longer actually change terminal color when run.
0.1.5
    Now works on Ubuntu.
0.1.4
    Implemented RESET_ALL on application exit
0.1.3
    Implemented init(wrap=False)
0.1.2
    Implemented init(autoreset=True)
0.1.1
    Minor tidy
0.1
    Works on Windows for foreground color, background color, bright or dim


.. |uumlaut| unicode:: U+00FC .. u with umlaut
   :trim:



old_spike.py

import sys
import re

class AnsiCode(object):
    RESET_ALL = 0
    BRIGHT = 1
    DIM = 2
    NORMAL = 22
    BLACK = 30
    RED = 31
    GREEN = 32
    YELLOW = 33
    BLUE = 34
    MAGENTA = 35
    CYAN = 36
    WHITE = 37
    DEFAULT = 39
    BLACK_BG = 40
    RED_BG = 41
    GREEN_BG = 42
    YELLOW_BG = 43
    BLUE_BG = 44
    MAGENTA_BG = 45
    CYAN_BG = 46
    WHITE_BG = 47
    DEFAULT_BG = 49

CSI = '\033['
ESCAPE = CSI + '%dm'

class Ansi(object):
    pass

for name in dir(AnsiCode):
    if not name.startswith('_'):
        code = getattr(AnsiCode, name)
        setattr(Ansi, name, ESCAPE % (code,))


class ColorStream(object):

    ANSI_RE = re.compile('\033\[(\d+)m')

    def __init__(self, wrapped, autoreset=False):
        self.wrapped = wrapped
        self.autoreset = autoreset
        self.enabled = sys.platform.startswith('win')

    def __getattr__( self, name ):
        return getattr( self.wrapped, name )

    def write( self, text ):
        if self.enabled:
            self.write_and_convert( text )
        else:
            self.wrapped.write( text )

        if self.autoreset:
            # but call win32 directly, don't convert this
            self.write_and_convert( Ansi.RESET_ALL )

    def write_and_convert(self, text):
        cursor = 0
        for match in self.ANSI_RE.finditer(text):
            # write regular text up to next ANSI sequence
            start, end = match.span()
            if cursor < start:
                snippet = text[cursor:start]
                self.wrapped.write( snippet )
            cursor = end

            # convert ANSI sequence to win32 calls
            ansi_code = int(match.group(1))
            if ansi_code in win32_calls:
                win32_calls[ansi_code]()
        if cursor < len(text) - 1:
            self.wrapped.write( text[cursor:] )



from ctypes import Structure, c_short, c_ushort, byref, windll

SHORT = c_short
WORD = c_ushort

class COORD(Structure):
  """struct in wincon.h."""
  _fields_ = [
    ("X", SHORT),
    ("Y", SHORT)]

class SMALL_RECT(Structure):
  """struct in wincon.h."""
  _fields_ = [
    ("Left", SHORT),
    ("Top", SHORT),
    ("Right", SHORT),
    ("Bottom", SHORT)]

class CONSOLE_SCREEN_BUFFER_INFO(Structure):
  """struct in wincon.h."""
  _fields_ = [
    ("dwSize", COORD),
    ("dwCursorPosition", COORD),
    ("wAttributes", WORD),
    ("srWindow", SMALL_RECT),
    ("dwMaximumWindowSize", COORD)]



# from wincon.h
BLACK     = 0x0000
BLUE      = 0x0001
GREEN     = 0x0002
CYAN      = 0x0003
RED       = 0x0004
MAGENTA   = 0x0005
YELLOW    = 0x0006
GREY      = 0x0007
INTENSE   = 0x0008

# from winbase.h
STD_OUTPUT_HANDLE= -11
HANDLE = windll.kernel32.GetStdHandle(STD_OUTPUT_HANDLE)


class MsWinTerm(object):

    def __init__(self):
        self._default_fore = GREY
        self._default_back = BLACK
        self.reset()


    def reset(self):
        self._fore = self._default_fore
        self._back = self._default_back
        self._intense_fore = True
        self._intense_back = False

    def reset_all(self):
        self.reset()
        self.set_text_attr()


    def fore(self, color=None):
        if color is None:
            color = self._default_fore
        self._fore = color
        self.set_text_attr()

    def back(self, color=None):
        if color is None:
            color = self._default_back
        self._back = color
        self.set_text_attr()

    def dim(self):
        self._intense_fore = False
        self._intense_back = False
        self.set_text_attr()

    def normal(self):
        self._intense_fore = True
        self._intense_back = False
        self.set_text_attr()

    def bright(self):
        self._intense_fore = True
        self._intense_back = True
        self.set_text_attr()


    @property
    def combined_attrs(self):
        fg = self._fore + self._intense_fore * INTENSE
        bg = self._back + self._intense_back * INTENSE
        return fg + bg * 16


    def set_text_attr(self):
        success = windll.kernel32.SetConsoleTextAttribute(
            HANDLE, self.combined_attrs)
        assert success


    def get_current_attr(self):
        """
        Returns the character attributes (colors) of the console screen buffer.
        """
        csbi = CONSOLE_SCREEN_BUFFER_INFO()
        windll.kernel32.GetConsoleScreenBufferInfo( HANDLE, byref(csbi) )
        return csbi.wAttributes

winterm = MsWinTerm()

win32_calls = {
    AnsiCode.RESET_ALL: winterm.reset_all,
    AnsiCode.BRIGHT: winterm.bright,
    AnsiCode.DIM: winterm.dim,
    AnsiCode.NORMAL: winterm.normal,
    AnsiCode.BLACK: lambda: winterm.fore(BLACK),
    AnsiCode.RED: lambda: winterm.fore(RED),
    AnsiCode.GREEN: lambda: winterm.fore(GREEN),
    AnsiCode.YELLOW: lambda: winterm.fore(YELLOW),
    AnsiCode.BLUE: lambda: winterm.fore(BLUE),
    AnsiCode.MAGENTA: lambda: winterm.fore(MAGENTA),
    AnsiCode.CYAN: lambda: winterm.fore(CYAN),
    AnsiCode.WHITE: lambda: winterm.fore(GREY),
    AnsiCode.DEFAULT: winterm.fore,
    AnsiCode.BLACK_BG: lambda: winterm.back(BLACK),
    AnsiCode.RED_BG: lambda: winterm.back(RED),
    AnsiCode.GREEN_BG: lambda: winterm.back(GREEN),
    AnsiCode.YELLOW_BG: lambda: winterm.back(YELLOW),
    AnsiCode.BLUE_BG: lambda: winterm.back(BLUE),
    AnsiCode.MAGENTA_BG: lambda: winterm.back(MAGENTA),
    AnsiCode.CYAN_BG: lambda: winterm.back(CYAN),
    AnsiCode.WHITE_BG: lambda: winterm.back(GREY),
    AnsiCode.DEFAULT_BG: winterm.back,
}

